GVFS/GVFS.Common/NuGetUpgrade/OrgNuGetUpgrader.cs

```diff
@@ -0,0 +1,244 @@
+                // https://github.com/dotnet/corefx/issues/20296
+                message = string.Format("Network error: could not connect to server ({0}). {1}", this.OrgInfoServerUrl, exception.Message);
+                this.TraceException(exception, nameof(this.TryQueryNewestVersion), $"Error connecting to server.");
+
+                return false;
+            }
+            catch (SerializationException exception)
+            {
+                message = string.Format("Parse error: could not parse response from server({0}). {1}", this.OrgInfoServerUrl, exception.Message);
+                this.TraceException(exception, nameof(this.TryQueryNewestVersion), $"Error parsing response from server.");
+
+                return false;
+            }
+
+            if (this.highestVersionAvailable != null &&
+                this.highestVersionAvailable > this.installedVersion)
+            {
+                newVersion = this.highestVersionAvailable;
+            }
+
+            if (newVersion != null)
+            {
+                this.tracer.RelatedInfo($"{nameof(this.TryQueryNewestVersion)} - new version available: installedVersion: {this.installedVersion}, highestVersionAvailable: {newVersion}");
+                message = $"New version {newVersion} is available.";
+                return true;
+            }
+            else if (this.highestVersionAvailable != null)
+            {
+                this.tracer.RelatedInfo($"{nameof(this.TryQueryNewestVersion)} - up-to-date");
+                message = $"highest version available is {this.highestVersionAvailable}, you are up-to-date";
+                return true;
+            }
+            else
+            {
+                this.tracer.RelatedInfo($"{nameof(this.TryQueryNewestVersion)} - no versions available from feed.");
+                message = $"No versions available via endpoint.";
+            }
+
+            newVersion = this.highestVersionAvailable;
+            message = null;
+            return true;
+        }
+
+        public class OrgNuGetUpgraderConfig : NuGetUpgraderConfig
+        {
+            public OrgNuGetUpgraderConfig(ITracer tracer, LocalGVFSConfig localGVFSConfig)
+                : base(tracer, localGVFSConfig)
+            {
+            }
+
+            public string OrgInfoServer { get; set; }
+
+            public string UpgradeRing { get; set; }
+
+            public override bool TryLoad(out string error)
+            {
+                if (!base.TryLoad(out error))
+                {
+                    return false;
+                }
+
+                if (!this.localConfig.TryGetConfig(GVFSConstants.LocalGVFSConfig.OrgInfoServerUrl, out string orgInfoServerUrl, out error))
+                {
+                    this.tracer.RelatedError(error);
+                    return false;
+                }
+
+                this.OrgInfoServer = orgInfoServerUrl;

```

Comment:
Can we have a method in the base class that is something like:
```
protected bool TrySetConfig(string key, Action<string> setValue, out string error)
{
	if (!this.localConfig.TryGetConfig(key, out string value, out error))
	{
		this.tracer.RelatedError(error);
		return false;
	}

	setValue(value);
}
```

Then this method could have something like:
```
   return base.TryLoad(out error) &&
        this.TrySetConfig(GVFSConstants.LocalGVFSConfig.OrgInfoServerUrl, x => this.OrgInfoServer = x, out error) &&
        this.TrySetConfig(GVFSConstants.LocalGVFSConfig.UpgradeRing, x => this.UpgradeRing= x, out error) &&
        etc...
```

The base class could do the same thing.

Reply:
I like this suggestion - but, I think I want to hold off on this change:

1) I want to make a bit larger change on this logic

2) This is not an area that we expect end users to configure directly. I want to make it more user friendly, but I think I can follow up with this work.